clippy_lints/src/extra_unused_type_parameters.rs

```diff
@@ -0,0 +1,175 @@
+            has_non_ty_params,
+            bound_map: FxHashMap::default(),
+            some_params_used: false,
+        }
+    }
+
+    fn emit_lint(&self) {
+        let (msg, help) = match self.map.len() {
+            0 => return,
+            1 => (
+                "type parameter goes unused in function definition",
+                "consider removing the parameter",
+            ),
+            _ => (
+                "type parameters go unused in function definition",
+                "consider removing the parameters",
+            ),
+        };
+
+        let source_map = self.cx.tcx.sess.source_map();
+        let span = if self.some_params_used || self.has_non_ty_params {
+            MultiSpan::from_spans(
+                self.map
+                    .iter()
+                    .map(|(def_id, &(mut span))| {
+                        if let Some(bound_span) = self.bound_map.get(def_id) {
+                            span = span.with_hi(bound_span.hi());
+                        }
+                        span = source_map.span_extend_to_next_char(span, ',', false);
+                        span = span.with_hi(BytePos(span.hi().0 + 1));
+
+                        let max_hi = self.generics.span.hi();
+                        if span.hi() >= max_hi {
+                            span = span.with_hi(BytePos(max_hi.0 - 1));
+                        }
+                        span
+                    })
+                    .collect(),
+            )
+        } else {
+            self.generics.span.into()
+        };
+
+        span_lint_and_help(self.cx, EXTRA_UNUSED_TYPE_PARAMETERS, span, msg, None, help);
+    }
+}
+
+impl<'cx, 'tcx> Visitor<'tcx> for TypeWalker<'cx, 'tcx> {
+    type NestedFilter = nested_filter::OnlyBodies;
+
+    fn visit_ty(&mut self, t: &'tcx Ty<'tcx>) {
+        if let Some((def_id, _)) = t.peel_refs().as_generic_param() {
+            if self.map.remove(&def_id).is_some() {
+                self.some_params_used = true;
+            }
+        } else if let TyKind::OpaqueDef(id, _, _) = t.kind {
+            // Explicitly walk OpaqueDef. Normally `walk_ty` would do the job, but it calls
+            // `visit_nested_item`, which checks that `Self::NestedFilter::INTER` is set. We're
+            // using `OnlyBodies`, so the check ends up failing and the type isn't fully walked.
+            let item = self.nested_visit_map().item(id);
+            walk_item(self, item);
+        } else {
+            walk_ty(self, t);
+        }
+    }
+
+    fn visit_where_predicate(&mut self, predicate: &'tcx WherePredicate<'tcx>) {
+        if let WherePredicate::BoundPredicate(predicate) = predicate {
+            // Collect spans for bounds that appear in the list of generics (not in a where-clause)
+            // for use in forming the help message
+            if let Some((def_id, _)) = predicate.bounded_ty.peel_refs().as_generic_param()
+                && predicate.span < self.generics.where_clause_span
```

Comment:
Huh, what does a comparison of spans do? :thinking:

Reply:
I assumed it's the same as doing a comparison on the tuple `(lo, hi)`, but I'm guessing comparing `.lo()` should work just as well.

Reply:
I worded my question badly: What is this span comparison trying to achieve?

Reply:
Here I'm looking for any bounds that appear inline with the list of generics, instead of in a where-clause. Comparing spans was the easiest way I thought of to rule out any bounds in the where-clause.

Reply:
A `WhereBoundPredicate` has a [`PredicateOrigin`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_hir/hir/enum.PredicateOrigin.html). Maybe you can use that instead?

Reply:
You're totally right.